src/lib.rs

```diff
@@ -143,78 +151,104 @@ impl SubnetActor for Actor {
-    //
-    //     st.mutate_state();
-    //     st.save();
-    //     Ok(None)
-    // }
-    //
+    /// Called by peers looking to leave a subnet.
+    fn leave<BS, RT>(rt: &mut RT) -> Result<Option<RawBytes>, ActorError>
+    where
+        BS: Blockstore,
+        RT: Runtime<BS>,
+    {
+        let caller = rt.message().caller();
+
+        // TODO: shall we check caller interface instead here?
+        // let code_cid = get_actor_code_cid(&caller).unwrap_or(Cid::default());
+        // if sdk::actor::get_builtin_actor_type(&code_cid) != Some(Type::Account) {
+        //     abort!(USR_FORBIDDEN, "caller not account actor type");
+        // }
+
+        let mut msg = None;
+        rt.transaction(|st: &mut State, rt| {
+            let stake = st.get_stake(rt.store(), &caller).map_err(|e| {
+                e.downcast_default(ExitCode::USR_ILLEGAL_STATE, "failed to load stake")
+            })?;
+
+            if stake.is_none() || stake.as_ref().is_some_and(|a| *a == TokenAmount::zero()) {
+                return Err(actor_error!(illegal_state, "caller has no stake in subnet"));
+            }
+
+            let stake = stake.unwrap();
+            if st.status != Status::Terminating {
+                msg = Some(CrossActorPayload::new(
+                    st.ipc_gateway_addr,
+                    ipc_gateway::Method::ReleaseStake as u64,
+                    RawBytes::serialize(FundParams {
+                        value: stake.clone(),
+                    })?,
+                    stake.clone(),
+                ));
+            }
+
+            // remove stake from balance table
+            st.rm_stake(&rt.store(), &caller, &stake).map_err(|e| {
+                e.downcast_default(ExitCode::USR_ILLEGAL_STATE, "cannot remove stake")
+            })?;
+
+            // now caller should have 0 balance now
+            st.mutate_state(&TokenAmount::zero());
+
+            Ok(true)
+        })?;
+
+        if let Some(p) = msg {
+            rt.send(p.to, p.method, p.params, p.value)?;
+        }
+
+        Ok(None)
+    }
+
+    fn kill<BS, RT>(rt: &mut RT) -> Result<Option<RawBytes>, ActorError>
+    where
+        BS: Blockstore,
+        RT: Runtime<BS>,
+    {
+        let mut msg = None;
+        rt.transaction(|st: &mut State, _| {
+            if st.status == Status::Terminating || st.status == Status::Killed {
+                return Err(actor_error!(
+                    illegal_state,
+                    "the subnet is already in a killed or terminating state"
+                ));
+            }
+
+            if !st.validator_set.is_empty() || st.total_stake != TokenAmount::zero() {
+                return Err(actor_error!(
+                    illegal_state,
+                    "this subnet can only be killed when all validators have left"
+                ));
+            }
+
+            // move to terminating state
+            st.status = Status::Terminating;
+
+            st.mutate_state(&TokenAmount::zero());
```

Comment:
Same here, we should use `TokenAmoun::zero()`, we should use `rt.curr_balance()`